magenta/interfaces/midi/midi.py

```diff
@@ -498,48 +516,216 @@ def stop_capture(self):
+      message: The incoming Control Change message.
+    """
+
+    # Metronome Velocity
+    if message.bytes()[:2] == _CC_MAP['Metronome velocity'].bytes()[:2]:
+      global _METRONOME_VELOCITY
+      _METRONOME_VELOCITY = message.bytes()[2]
+
+class CCRemapper(object):
+  """CC Message Remapping interface.
+
+  Attributes:
+    _inport: The Mido port for receiving messages.
+    _cc_map: The dictionary of user defined Control Change messages.
+
+  Args:
+    input_midi_port: The string MIDI port name to use for input.
+  """
+
+  def __init__(self, input_midi_port):
+    self._inport = mido.open_input(input_midi_port)
+    self._cc_map = _CC_MAP
+
+  def remapper_interface(self):
+    """Interface asking for user input of which parameters to remap"""
+
+    while True:
+      print "\nWhat would you like to assign a Control Change to...?:"
+      print "0. None (Exit)"
+      for i, CCparam in enumerate(sorted(self._cc_map.keys())):
+        print "{}. {}".format(i + 1, CCparam)
+      print ""
+      try:
+        cc_choice_index = int(raw_input("Please make a selection...\n>>>"))
+      except ValueError:
+        print "Please enter a number..."
+        time.sleep(1)
+        continue
+
+      else:
+        if cc_choice_index == 0:
+          self._inport.close()
+          return self._cc_map
+        elif cc_choice_index < 0 or cc_choice_index > len(self._cc_map):
+          print("There is no CC Parameter assigned to that "
+                "number, please select from the list.")
+          time.sleep(1)
+          continue
+        else:
+          cc_choice = sorted(self._cc_map.keys())[cc_choice_index - 1]
+
+      if cc_choice == 'Start Capture' or cc_choice == 'Stop Capture':
+        self.remap_capture_message(self._inport, cc_choice,
+                                   self._cc_map['Start Capture'],
+                                   self._cc_map['Stop Capture'])
+        time.sleep(1)
+      else:
+        self.remap_cc_message(self._inport, cc_choice)
+        time.sleep(1)
+
+  def remap_cc_message(self, input_port, cc_choice):
+    """Defines how to remap control change messages for defined parameters.
+
+    Args:
+      input_port: The input port to receive the control change message.
+      cc_choice: The CC map dictionary key to assign a message to.
+    """
+
+    while True:
+      while input_port.receive(block=False) is not None:
+        pass
+      print ("What control or key would you like to assign to {}?\n"
+             "Please press one now...".format(cc_choice))
+      msg = input_port.receive()
+      if msg.hex().startswith('B'):
+        print ("{} has been assigned to controller "
+               "number {}".format(cc_choice, msg.control))
+        self._cc_map[cc_choice] = msg
+        return
+      else:
+        print('Sorry, I only accept MIDI CC messages for this parameter..')
+        time.sleep(.5)
+        continue
+
+  def remap_capture_message(self, input_port, cc_choice, msg_start_capture,
```

Comment:
It is the only CC that has a toggleability (for use with basic foot switches, notes on the keyboard, or MIDI buttons that only send out 1 message). So it needs a separate set of assignment definitions

> On Sep 12, 2016, at 9:53 AM, Adam Roberts <notifications@github.com> wrote:
> 
> In magenta/interfaces/midi/midi.py <https://github.com/tensorflow/magenta/pull/209#discussion_r78410295>:
> 
> > +      while input_port.receive(block=False) is not None:
> > +        pass
> > +      print ("What control or key would you like to assign to {}?\n"
> > +             "Please press one now...".format(cc_choice))
> > +      msg = input_port.receive()
> > +      if msg.hex().startswith('B'):
> > +        print ("{} has been assigned to controller "
> > +               "number {}".format(cc_choice, msg.control))
> > +        self._cc_map[cc_choice] = msg
> > +        return
> > +      else:
> > +        print('Sorry, I only accept MIDI CC messages for this parameter..')
> > +        time.sleep(.5)
> > +        continue
> > +
> > +  def remap_capture_message(self, input_port, cc_choice, msg_start_capture,
> Why do we need something specific for the capture message?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/tensorflow/magenta/pull/209/files/23b775fe695f914449d738c4e11817a65587fc8f#r78410295>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AP7D9U4rs4Blz__WDBTKcPhiDkHXJ41Xks5qpYOngaJpZM4J6Nk5>.
>